File: 62. Unique Paths.py
class Solution(object):
    def countOfPath(self, m, n, i, j, cnt):
        """
        type i: int current row position
        type j: int current col position
        type cnt: int current path count to point(i, j)
        
        """
        curcnt = cnt
        if i != 0 and j != 0:
            return self.countOfPath(m, n, i - 1, j, curcnt) + self.countOfPath(m, n, i, j - 1, curcnt)
        elif i == 0 and j != 0:
            return self.countOfPath(m, n, i, j - 1, curcnt)
        elif i != 0 and j == 0:
            return self.countOfPath(m, n, i - 1, j, curcnt)
        if i == 0 and j == 0:
            return cnt
        
    def uniquePaths(self, m, n):
        """
        :type m: int
        :type n: int
        :rtype: int
        """
        cnt = self.countOfPath(m, n, m - 1, n - 1, 0)
        return cnt
        

File: 62. Unique Paths.c
int uniquePaths(int m, int n) {
    int grid[m][n];
    int i, j;
    grid[0][0] = 0;
    if (m > 1 && n > 1){
        for (i = 1; i < m; i++)
            grid[i][0] = 1;
        for (j = 1; j < n; j++)
            grid[0][j] = 1;
        for (i = 1; i < m; i++)
            for (j = 1; j < n; j++)
                grid[i][j] = grid[i - 1][j] + grid[i][j - 1];
        return grid[m - 1][n - 1];
    }else 
        return !(!m || !n); // m or n = 0, no path; m or n = 1, only 1 path   
}
